### DISCUSSION TITLE
simulateTouch 怎么在注入 Springboard的时候点击我app呢

### DISCUSSION THREAD

**Discussion by lengfeng222** [community/community#127426] [Programming Help]
### Select Topic Area

Product Feedback

### Body

#import "TouchSimulator.h"
#import <UIKit/UIKit.h>
#import <dlfcn.h>
#import <mach/mach_time.h>

static void postEvent(IOHIDEventRef event);
static void execute();
static IOHIDEventRef parent = NULL;

void simulateTouch(int type, float x, float y) {
    NSLog(@"冷风牛逼 - simulateTouch - 类型: %d, X: %f, Y: %f", type, x, y);

    if (parent == NULL) {
        parent = IOHIDEventCreateDigitizerEvent(kCFAllocatorDefault,
                                                mach_absolute_time(),
                                                kIOHIDDigitizerTransducerTypeHand,
                                                0,
                                                0,
                                                kIOHIDDigitizerEventTouch,
                                                0,
                                                0.0,
                                                0.0,
                                                0.0,
                                                0.0,
                                                0.0,
                                                0,
                                                true,
                                                0
                                                );
        IOHIDEventSetIntegerValue(parent, kIOHIDEventFieldDigitizerIsDisplayIntegrated, 1);
        NSLog(@"冷风牛逼 - 创建父事件");
    }

    uint32_t isTouch = type == TOUCH_UP ? 0 : 1;
    IOHIDDigitizerEventMask eventMask = 0;
    if (type != TOUCH_UP && type != TOUCH_DOWN) {
        eventMask |= kIOHIDDigitizerEventPosition;
    }
    if (type == TOUCH_UP || type == TOUCH_DOWN) {
        eventMask |= (kIOHIDDigitizerEventTouch | kIOHIDDigitizerEventRange);
    }

    IOHIDEventRef child = IOHIDEventCreateDigitizerFingerEvent(kCFAllocatorDefault, mach_absolute_time(), 1, 3, eventMask, x, y, 0.0f, 0.0f, 0.0f, isTouch, isTouch, 0);
    IOHIDEventSetFloatValue(child, kIOHIDEventFieldDigitizerMajorRadius, 0.04f);
    IOHIDEventSetFloatValue(child, kIOHIDEventFieldDigitizerMinorRadius, 0.04f);
    IOHIDEventAppendEvent(parent, child);
    NSLog(@"冷风牛逼 - 创建并附加子事件");

    execute();
}

static UIWindow* getKeyWindow() {
    for (UIWindow *window in [UIApplication sharedApplication].windows) {
        if (window.isKeyWindow) {
            return window;
        }
    }
    return NULL;
}

static void postEvent(IOHIDEventRef event) {
    static IOHIDEventSystemClientRef ioSystemClient = nil;
    if (ioSystemClient == NULL) {
        ioSystemClient = IOHIDEventSystemClientCreate(kCFAllocatorDefault);
        NSLog(@"冷风牛逼 - 创建 IOHIDEventSystemClient");
    }
    if (event != NULL) {
        IOHIDEventSystemClientDispatchEvent(ioSystemClient, event);
        NSLog(@"冷风牛逼 - 事件已发送");
    }
}

static void execute() {
    UIWindow* keyWindow = getKeyWindow();
    if (keyWindow != NULL) {
        NSLog(@"冷风牛逼 - 获取到主窗口: %@", keyWindow);
    } else {
        NSLog(@"冷风牛逼 - 未能获取到主窗口");
    }

    IOHIDEventSetIntegerValue(parent, kIOHIDEventFieldDigitizerTiltX, kIOHIDDigitizerTransducerTypeHand);
    IOHIDEventSetIntegerValue(parent, kIOHIDEventFieldDigitizerTiltY, 1);
    IOHIDEventSetIntegerValue(parent, kIOHIDEventFieldDigitizerAltitude, 1);

    if (keyWindow != NULL) {
        postEvent(parent);
    } else {
        NSLog(@"冷风牛逼 - 无法发送事件，因为未能获取到主窗口");
    }

    CFRelease(parent);
    parent = NULL;
    NSLog(@"冷风牛逼 - 事件已执行");
}

void simulateSwipe(float startX, float startY, float endX, float endY, int steps) {
    float stepX = (endX - startX) / steps;
    float stepY = (endY - startY) / steps;

    NSLog(@"冷风牛逼 - simulateSwipe - 起始X: %f, 起始Y: %f, 结束X: %f, 结束Y: %f, 步数: %d", startX, startY, endX, endY, steps);

    // 在起始位置触摸
    simulateTouch(TOUCH_DOWN, startX, startY);

    // 小步移动到结束位置
    for (int i = 1; i <= steps; i++) {
        simulateTouch(TOUCH_MOVE, startX + stepX * i, startY + stepY * i);
        usleep(10000); // 模拟真实滑动的短暂延迟
    }

    // 在结束位置抬起
    simulateTouch(TOUCH_UP, endX, endY);
}

**Comment by ebndev** [comment]
Hi @lengfeng222, thanks for participating!

Unfortunately, we don’t currently have moderators for languages other than English. Until that changes, we need to ask that everyone use English here in the GitHub Community when posting to ensure discussions are adhering to our [Code of Conduct](https://github.com/community/community/blob/920a928ffe3ae4863b64a4ea26917046e11f5c1a/CODE_OF_CONDUCT.md). 

We’ll be locking this post for now, but please feel encouraged to create a new discussion in English if you still need support.
